kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h

```diff
@@ -79,69 +79,80 @@
+{
+  lv_32fc_t diff;
+  for(unsigned int i = 0; i < num_points; ++i) {
+    /*
+     * Calculate: |y - x|^2 * SNR_lin
+     * Compare C++: *target++ = scalar * std::norm(symbol - *constellation++);
+     */
+    diff = symbol - *points++;
+    *target++ = scalar * (lv_creal(diff) * lv_creal(diff) + lv_cimag(diff) * lv_cimag(diff));
+  }
+}
 
 
 #ifdef LV_HAVE_AVX2
 #include<immintrin.h>
 
 static inline void
-volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, lv_32fc_t* points,
-                                                     float scalar, unsigned int num_points)
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, 
+                                                     lv_32fc_t* points, float scalar, 
+                                                     unsigned int num_points)
 {
   const unsigned int num_bytes = num_points*8;
-  __m128 xmm0, xmm9, xmm10, xmm11;
-  __m256 xmm1, xmm2, xmm3, xmm4, xmm5, xmm6, xmm7, xmm8;
+  __m128 xmm9, xmm10;
+  __m256 xmm4, xmm5, xmm6, xmm7;
+  __m256 xmm_points0, xmm_points1, xmm_result;
 
-  lv_32fc_t diff;
-  memset(&diff, 0x0, 2*sizeof(float));
-
-  float sq_dist = 0.0;
   int bound = num_bytes >> 6;
   int leftovers0 = (num_bytes >> 5) & 1;
   int leftovers1 = (num_bytes >> 4) & 1;
   int leftovers2 = (num_bytes >> 3) & 1;
-  int i = 0;
-
-  __m256i idx = _mm256_set_epi32(7,6,3,2,5,4,1,0);
-  xmm1 = _mm256_setzero_ps();
-  xmm2 = _mm256_load_ps((float*)&points[0]);
-  xmm8 = _mm256_set1_ps(scalar);
-  xmm11 = _mm256_extractf128_ps(xmm8,1);
-  xmm0 = _mm_load_ps((float*)src0);
-  xmm0 = _mm_permute_ps(xmm0, 0b01000100);
-  xmm1 = _mm256_insertf128_ps(xmm1, xmm0, 0);
-  xmm1 = _mm256_insertf128_ps(xmm1, xmm0, 1);
-  xmm3 = _mm256_load_ps((float*)&points[4]);
-
-  for(; i < bound; ++i) {
-    xmm4 = _mm256_sub_ps(xmm1, xmm2);
-    xmm5 = _mm256_sub_ps(xmm1, xmm3);
+  
+  // load complex value into all parts of the register.
+  const __m256 xmm_symbol = _mm256_castpd_ps(_mm256_broadcast_sd((const double*)src0));
+  const __m128 xmm128_symbol = _mm256_extractf128_ps(xmm_symbol, 1);
+  
+  // Load scalar into all 8 parts of the register
+  const __m256 xmm_scalar = _mm256_broadcast_ss(&scalar);
+  const __m128 xmm128_scalar = _mm256_extractf128_ps(xmm_scalar, 1);
+
+  // Set permutation constant
+  const __m256i idx = _mm256_set_epi32(7,6,3,2,5,4,1,0);
+  
+  xmm_points0 = _mm256_load_ps((float*)points);
+  xmm_points1 = _mm256_load_ps((float*)(points + 4));
+
+  for(int i = 0; i < bound; ++i) {
+    xmm4 = _mm256_sub_ps(xmm_symbol, xmm_points0);
+    xmm5 = _mm256_sub_ps(xmm_symbol, xmm_points1);
     points += 8;
     xmm6 = _mm256_mul_ps(xmm4, xmm4);
     xmm7 = _mm256_mul_ps(xmm5, xmm5);
 
-    xmm2 = _mm256_load_ps((float*)&points[0]);
+    xmm_points0 = _mm256_load_ps((float*)points);
+    xmm_points1 = _mm256_load_ps((float*)(points + 4));
 
     xmm4 = _mm256_hadd_ps(xmm6, xmm7);
     xmm4 = _mm256_permutevar8x32_ps(xmm4, idx);
```

Comment:
Just a rant: At first sight I stumbled a little and wondered why this is required. But than I remembered what a pain and how broken AVX is if you deal with complex numbers. I really envy NEON for having ``vld2``.

Reply:
`_mm256_permutevar8x32_ps` is actually the reason the AVX2 kernel is faster than the AVX one. Before that intrinsic was added, there was a good chance you end up with an AVX kernel that is slower than its SSE counterpart.

Reply:
I am aware of that. But I better not digress any further…